agent/src/flow_generator/protocol_logs/mq/kafka.rs

```diff
@@ -352,22 +452,125 @@ impl KafkaLog {
+                if payload.len() < KAFKA_RESP_HEADER_LEN {
+                    return Err(Error::KafkaLogParseFailed);
+                }
                 self.response(payload, info)?;
                 self.perf_stats.as_mut().map(|p| p.inc_resp());
             }
         }
         Ok(())
     }
 
+    fn get_topics_name(
+        &mut self,
+        api_key: u16,
+        api_version: u16,
+        payload: &[u8],
+    ) -> Option<String> {
+        let mut _fixed_offset = kafka_apiversion_topic_fixed_offset!(api_key, api_version);
+        if _fixed_offset == usize::max_value() {
+            return None;
+        }
+        if _fixed_offset > payload.len() {
+            return None;
+        }
+        match api_key {
+            KAFKA_PRODUCE => {
+                if api_version >= 3 && api_version <= 8 {
+                    let tid_len = read_i16_be(&payload[0..2]);
+                    if tid_len > 0 {
+                        _fixed_offset = _fixed_offset + tid_len as usize
+                    }
+                }
+                // Significant improvements since version 9
+                if api_version == 9 {
+                    let tid_len = payload[0];
+                    if tid_len > 0 {
+                        _fixed_offset = _fixed_offset + tid_len as usize
+                    }
+                    if _fixed_offset + 1 + payload[_fixed_offset] as usize > payload.len() {
+                        return None;
+                    }
+                    let (_, _, result) = UTF_16BE.decode(
+                        &payload
+                            [_fixed_offset.._fixed_offset + 1 + payload[_fixed_offset] as usize],
+                    );
+                    if result {
+                        return Some(
+                            String::from_utf8_lossy(
+                                &payload[_fixed_offset
+                                    .._fixed_offset + 1 + payload[_fixed_offset] as usize],
+                            )
+                            .into_owned(),
+                        );
+                    } else {
+                        return None;
+                    };
+                }
+                if _fixed_offset + 2 > payload.len() {
+                    return None;
+                }
+                let len = read_u16_be(&payload[_fixed_offset.._fixed_offset + 2]);
+                if _fixed_offset + 2 + len as usize > payload.len() {
+                    return None;
+                }
+                return Some(
+                    String::from_utf8_lossy(
+                        &payload[_fixed_offset + 2.._fixed_offset + 2 + len as usize],
+                    )
+                    .into_owned(),
+                );
+            }
+            KAFKA_FETCH => {
+                // The 12th version is a transitional version, and the decoding protocols of the previous versions are more different.
+                if api_version == 12 {
```

Comment:
Same issue here. Use match guard for different versions.

Reply:
Same answer